File: algorithms/python/mergesort.py
import sys
import time

def merge(arr, tmp, l, m, r):
    tmp[l:r+1] = arr[l:r+1]
    i, j, k = l, m + 1, l
    while i <= m and j <= r:
        if tmp[i] <= tmp[j]:
            arr[k] = tmp[i]; i += 1
        else:
            arr[k] = tmp[j]; j += 1
        k += 1
    while i <= m:
        arr[k] = tmp[i]; i += 1; k += 1
    while j <= r:
        arr[k] = tmp[j]; j += 1; k += 1

def mergesort(arr, tmp, l, r):
    if l >= r:
        return
    m = l + (r - l) // 2
    mergesort(arr, tmp, l, m)
    mergesort(arr, tmp, m + 1, r)
    merge(arr, tmp, l, m, r)

def main():
    data = sys.stdin.read().split()
    n    = int(data[0])
    arr  = list(map(int, data[1:n+1]))
    tmp  = [0] * n

    # افزایش recursion limit برای n بزرگ
    sys.setrecursionlimit(n * 2 + 100)

    start  = time.perf_counter()
    mergesort(arr, tmp, 0, n - 1)
    ms = (time.perf_counter() - start) * 1000

    sys.stdout.write(" ".join(map(str, arr)) + "\n")
    sys.stderr.write(f"{ms:.3f}\n")

if __name__ == "__main__":
    main()

File: algorithms/julia/mergesort.jl
function merge!(arr::Vector{Int64}, tmp::Vector{Int64}, l::Int, m::Int, r::Int)
    @inbounds tmp[l:r] .= arr[l:r]
    i, j, k = l, m + 1, l
    @inbounds while i <= m && j <= r
        if tmp[i] <= tmp[j]
            arr[k] = tmp[i]; i += 1
        else
            arr[k] = tmp[j]; j += 1
        end
        k += 1
    end
    @inbounds while i <= m; arr[k] = tmp[i]; i += 1; k += 1; end
    @inbounds while j <= r; arr[k] = tmp[j]; j += 1; k += 1; end
end

function mergesort!(arr::Vector{Int64}, tmp::Vector{Int64}, l::Int, r::Int)
    l >= r && return
    m = l + (r - l) ÷ 2
    mergesort!(arr, tmp, l, m)
    mergesort!(arr, tmp, m + 1, r)
    merge!(arr, tmp, l, m, r)
end

function main()
    input = split(read(stdin, String))
    n     = parse(Int, input[1])
    arr   = parse.(Int64, input[2:n+1])
    tmp   = similar(arr)

    # JIT warmup
    w = copy(arr[1:min(10, n)])
    wt = similar(w)
    mergesort!(w, wt, 1, length(w))

    start = time_ns()
    mergesort!(arr, tmp, 1, n)
    ms = (time_ns() - start) / 1e6

    println(join(arr, " "))
    println(stderr, ms)
end

main()